akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedBehavior.scala

```diff
@@ -0,0 +1,113 @@
+  @InternalApi private[akka] final case class ReceiveTimeout(timeout: akka.actor.ReceiveTimeout) extends EventsourcedProtocol
+
+  /** INTERNAL DSL */
+  private[akka] implicit class UnlessStopped[T](val b: Behavior[T]) extends AnyVal {
+    /** Returns `other` behavior if the wrapped one is `Stopped`. */
+    def unlessStopped(other: ⇒ Behavior[T]): Behavior[T] =
+      b match {
+        case _: StoppedBehavior[T @unchecked] ⇒ b // still the stopped one
+        case _                                ⇒ other
+      }
+  }
+
+  implicit object PersistentBehaviorLogSource extends LogSource[EventsourcedBehavior[_, _, _]] {
+    override def genString(b: EventsourcedBehavior[_, _, _]): String = {
+      val behaviorShortName = b match {
+        case _: EventsourcedRunning[_, _, _]                  ⇒ "running"
+        case _: EventsourcedRecoveringEvents[_, _, _]         ⇒ "recover-evts"
+        case _: EventsourcedRecoveringSnapshot[_, _, _]       ⇒ "recover-snap"
+        case _: EventsourcedRequestingRecoveryPermit[_, _, _] ⇒ "awaiting-permit"
+      }
+      s"PersistentBehavior[id:${b.persistenceId}][${b.context.self.path}][$behaviorShortName]"
+    }
+  }
+
+}
+
+trait EventsourcedBehavior[Command, Event, State] {
+  import EventsourcedBehavior._
+  import akka.actor.typed.scaladsl.adapter._
+
+  def context: ActorContext[Any]
+  def timers: TimerScheduler[Any]
+
+  type C = Command
+  type AC = ActorContext[C]
+  type E = Event
+  type S = State
+
+  // used for signaling intent in type signatures
+  type SeqNr = Long
+
+  // FIXME make this not even visible in the state that is "running", we'll never call it there again
+  def persistenceId: String
+
+  def initialState: State
+  def commandHandler: PersistentBehaviors.CommandHandler[Command, Event, State]
+  def eventHandler: (State, Event) ⇒ State
+
+  // FIXME make this not even visible in the state that is "running", we'll never call it there again
+  def recoveryCompleted: (ActorContext[Command], State) ⇒ Unit
+  def snapshotWhen: (State, Event, SeqNr) ⇒ Boolean
+
+  def tagger: Event ⇒ Set[String]
+  def journalPluginId: String
+  def snapshotPluginId: String
+
+  // ------ common -------
+
+  // FIXME make all things protected
+
+  lazy val extension = Persistence(context.system.toUntyped)
+  lazy val journal: a.ActorRef = extension.journalFor(journalPluginId)
+  lazy val snapshotStore: a.ActorRef = extension.snapshotStoreFor(snapshotPluginId)
+
+  protected lazy val selfUntyped: a.ActorRef = context.self.toUntyped
+  protected lazy val selfUntypedAdapted: a.ActorRef = context.messageAdapter[Any] {
+    case res: JournalProtocol.Response           ⇒ JournalResponse(res)
+    case RecoveryPermitter.RecoveryPermitGranted ⇒ RecoveryPermitGranted
+    case res: SnapshotProtocol.Response          ⇒ SnapshotterResponse(res)
+    case cmd: Command @unchecked                 ⇒ cmd // if it was wrong, we'll realise when trying to onMessage the cmd
+  }.toUntyped
```

Comment:
We definitely need the undadapted one, since people watch it. See the intial behaviors